tests/unit/cluster/slot-migration.tcl

```diff
@@ -0,0 +1,363 @@
+proc get_open_slots {srv_idx} {
+    foreach line [split [R $srv_idx CLUSTER NODES] "\n"] {
+        set line [string trim $line]
+        if {$line eq {}} continue
+        if {[regexp {myself} $line] == 0} continue
+        set slots {}
+        regexp {\[.*} $line slots
+        return $slots
+    }
+}
+
+proc get_cluster_role {srv_idx} {
+    foreach line [split [R $srv_idx CLUSTER NODES] "\n"] {
+        set line [string trim $line]
+        if {$line eq {}} continue
+        if {[regexp {myself} $line] == 0} continue
+        set role {}
+        regexp {myself,(\w+)} $line -> role
+        return $role
+    }
+}
+
+proc wait_for_role {srv_idx role} {
+    set node_timeout [lindex [R 0 CONFIG GET cluster-node-timeout] 1]
+    # wait for a gossip cycle for states to be propagated throughout the cluster
+    after $node_timeout
+    wait_for_condition 100 100 {
+        [lindex [split [R $srv_idx ROLE] " "] 0] eq $role
+    } else {
+        fail "R $srv_idx didn't assume the replication $role in time"
+    }
+    wait_for_condition 100 100 {
+        [get_cluster_role $srv_idx] eq $role
+    } else {
+        fail "R $srv_idx didn't assume the cluster $role in time"
+    }
+    wait_for_cluster_propagation
+}
+
+proc wait_for_slot_state {srv_idx pattern} {
+    wait_for_condition 100 100 {
+        [get_open_slots $srv_idx] eq $pattern
+    } else {
+        fail "incorrect slot state on R $srv_idx: expected $pattern; got [get_open_slots $srv_idx]"
+    }
+}
+
+# Check if the server responds with "PONG"
+proc check_server_response {server_id} {
+    # Send a PING command and check if the response is "PONG"
+    return [expr {[catch {R $server_id PING} result] == 0 && $result eq "PONG"}]
+}
+
+# restart a server and wait for it to come back online
+proc restart_server_and_wait {server_id} {
```

Comment:
Almost universally we de-schedule processes by using the pause helpers to trigger failovers. This does two things, one is that we don't have to restart the process manually later, and second is that it's similar to when a processes disappears for a bit so tests additional modes.

Is there a specific reason we are restarting here?

Reply:
>Almost universally we de-schedule processes by using the pause helpers to trigger failovers.

is this `pause_process`? it kills the process and a restart is needed too. or this is something else?

Reply:
It just pauses the process, I was asking if we specifically need to kill it here?

Reply:
I see. my intent is to simulate true failures. If I don't kill the server, I will have to use "graceful failover" to switch roles.